File: letteamnamestringnil/letteamnamestringnil/CharacterSheet.swift
//
//  Player.swift
//  letteamnamestringnil
//
//  Created by Alexis Covacha on 10/31/21.
//

import Foundation
import SwiftUI

class CharacterSheet: ObservableObject {

    @Published var basicInfo: PlayerBasicInfo
    @Published var stats: PlayerStats
    @Published var skills: Skills
    @Published var attacks: String
    @Published var spellcasting: [String: Any]
    @Published var spells: [Spells]
    @Published var inventory: PlayerInventory
    @Published var background: PlayerBackgroundInfo
    @Published var notes: String = ""

//    @Published var hitDice: [Int: Int]

    init(json: [String: Any]) {

        var character: [String: Any] = [:]
        character["name"]      = json["name"]
        character["class"]     = json["class"]
        character["race"]      = json["race"]
        character["alignment"] = json["alignment"]
        character["age"] = json["age"]
        character["eyes"] = json["eyes"]
        character["height"] = json["height"]
        character["weight"] = json["weight"]
        character["isNPC"]      = json["is_npc"]
        character["isFriendly"] = json["is_friendly"]

        var stats: [String: Any] = json["stats"] as! [String: Any]
        stats["ac"]            = json["armor_class"]
        stats["initiative"]    = json["initiative"]
        stats["maxHP"]         = json["max_hp"]
        stats["curHP"]         = json["cur_hp"]
        stats["tmpHP"]         = json["tmp_hp"]
        stats["hitDice"]       = json["hit_dice"]
        stats["level"]         = json["level"]
        stats["xp"]            = json["experience_points"]
        stats["speed"]         = json["speed"]
        stats["abilityScores"] = json["abilities"]
        stats["perception"]    = json["passive_wisdom"]
        stats["inspiration"]   = json["inspiration"]
        stats["profBonus"]     = json["proficiency_bonus"]
        stats["savingThrows"]  = json["saving"]
        stats["deathSaveSuccess"] = json["death_save_success"]
        stats["deathSaveFailure"] = json["death_save_failure"]

        let personality = Personality(json: json)
        let spellsData = (json["spells"] as! [String: Any])["by_level"] as! [[String: Any]]

        self.basicInfo = PlayerBasicInfo(characterInfo: character)
        self.stats = PlayerStats(stats: stats)
        self.skills = Skills(sk: json["skills"] as! [String: Bool])
        self.attacks = json["attacks"] as! String
        self.inventory = PlayerInventory(equipment: json["equipment"] as? String ?? "",
                                         darkGifts: json["dark_gifts"] as? String ?? "",
                                         treasure: json["treasure"] as? String ?? "")
        self.background = PlayerBackgroundInfo(
            background: json["background"] as! String,
            backstory: json["backstory"] as! String,
            proficiencies: json["other_proficiencies_Skills"] as! String,
            features: json["features"] as! String,
            featuresAndTraits: json["features_and_traits"] as! String,
                allies: json["allies"] as! String,
            personality: personality,
            appearance: json["appearance"] as! [String: Any]
        )
        self.notes = json["notes"] as! String
        self.spellcasting = json["spellcasting"] as! [String: Any]
        self.spells = []

        for level in spellsData {
            for spell in level["spells"] as! [[String: Any]] {
                self.spells.append(Spells(name: spell["name"] as! String,
                        charges: spell["charges"] as! Int))
            }
        }
    }

    func toDictionary() -> [String:Any] {
        // TODO IMPLEMENT
        return [:]
    }

}

struct PlayerBasicInfo: Hashable {
    var name: String
    var className: String
    var isNPC: Bool
    var isFriendly: Bool
    var race: String
    var alignment: Int

    init(characterInfo: [String: Any]) {
        self.name      = characterInfo["name"] as! String
        self.className = characterInfo["clss"] as! String
        self.race      = characterInfo["race"] as! String
        self.alignment = characterInfo["alignment"] as! Int
        self.isNPC = characterInfo["isNPC"] as! Bool
        self.isFriendly = characterInfo["isFriendly"] as! Bool
    }
}

struct PlayerStats: Hashable {
    var ac: Int
    var initiative: Int
    var maxHP, curHP, tmpHP: Int
    var hitDice: Int
    var level: Int
    var xp: Int
    var speed: Int
    var abilityScores: [String:Int]
    var perception: Int
    var inspiration: Int
    var profBonus: Int
    var savingThrows: SavingThrows
    var deathSaveSuccess: Int
    var deathSaveFailure: Int

    init(stats: [String: Any]) {
        ac = stats["ac"] as! Int
        initiative = stats["initiative"] as! Int
        maxHP = stats["maxHP"] as! Int
        curHP = stats["curHP"] as! Int
        tmpHP = stats["tmpHP"] as! Int
        hitDice = stats["hitDice"] as! Int
        level = stats["level"] as! Int
        xp = stats["xp"] as! Int
        speed = stats["speed"] as! Int
        abilityScores = stats["abilityScores"] as! [String:Int]
        perception = stats["perception"] as! Int
        inspiration = stats["inspiration"] as! Int
        profBonus = stats["profBonus"] as! Int
        savingThrows = SavingThrows(st: stats["savingThrows"] as! [String:Bool])
        deathSaveSuccess = stats["deathSaveSuccess"] as! Int
        deathSaveFailure = stats["deathSaveFailure"] as! Int
    }

}

struct Spells: Hashable {
    var name: String
    var charges: Int
}


struct PlayerBackgroundInfo {
    var background: String
    var backstory: String
    var proficiencies: String
    var features: String
    var featuresAndTraits: String
    var allies: String
    var personality: Personality
    var appearance: [String: Any]
}

struct Personality: Hashable {
    var traits: String?
    var ideals: String?
    var bonds: String?
    var flaws: String?

    init(json: [String: Any]) {
        traits = json["traits"] as? String
        ideals = json["ideals"] as? String
        bonds  = json["bonds"] as? String
        flaws  = json["flaws"] as? String
    }
}

struct PlayerInventory: Hashable {
    var equipment: String
    var darkGifts: String
    var treasure: String
}

struct SavingThrows: Hashable {
    var str, dex, con, int, wis, char: Bool

    init(st: [String: Bool]) {
        str = st["strength"]!
        dex = st["dexterity"]!
        con = st["constitution"]!
        int = st["intelligence"]!
        wis = st["wisdom"]!
        char = st["charisma"]!
    }
}

struct Skills: Hashable {
    var acrobatics, animalHandling, arcana, athletics, deception, history, insight,
        intimidation, investigation, medicine, nature, perception, performance, persuasion,
        religion, sleightOfHand, stealth, survival: Bool

    init(sk: [String: Bool]) {
        acrobatics = sk["acrobatics"]!
        animalHandling = sk["animal_handling"]!
        arcana = sk["arcana"]!
        athletics = sk["athletics"]!
        deception = sk["deception"]!
        history = sk["history"]!
        insight = sk["insight"]!
        intimidation = sk["intimidation"]!
        investigation = sk["investigation"]!
        medicine = sk["medicine"]!
        nature = sk["nature"]!
        perception = sk["perception"]!
        performance = sk["performance"]!
        persuasion = sk["persuasion"]!
        religion = sk["religion"]!
        sleightOfHand = sk["sleight_of_hand"]!
        stealth = sk["stealth"]!
        survival = sk["survival"]!
    }
}

File: backend/django/app/charactersheet.py
import logging
from django.db import transaction
from jsonschema import validate, ValidationError
import app.models


logger = logging.getLogger(__name__)

same_attrs = ["name", "clss", "is_npc", "is_friendly", "level", "race", "background", "alignment", "experience_points", "inspiration", "proficiency_bonus", "passive_wisdom", "other_proficiencies_skills", "armor_class", "initiative", "speed", "max_hp", "cur_hp", "temp_hp", "hit_dice", "death_save_success", "death_save_failure", "attacks", "equipment", "traits", "ideals", "bonds", "flaws", "features_and_traits", "age", "eyes", "height", "weight", "backstory", "dark_gifts", "features", "allies", "treasure", "notes"]

abilities = ["strength", "dexterity", "constitution", "intelligence", "wisdom", "charisma"]

skills = ["acrobatics", "animal_handling", "arcana", "athletics", "deception", "history", "insight", "intimidation", "investigation", "medicine", "nature", "perception", "performance", "persuasion", "religion", "sleight_of_hand", "stealth", "survival"]

spells_same_attrs = ["spellcast_ability", "spell_save_dc", "spell_attack_bonus"]

spell_same_attrs = ["name", "description", "classes", "level", "components", "material", "casting_time", "die_sides", "die_count", "extra_damage", "damage_type", "duration", "range", "school", "target"]

character_sheet_schema = {
    "type": "object",
    "properties": {
        "name": {"type": "string", "maxLength": 255},
        "clss": {"type": "string", "maxLength": 255},
        "is_npc": {"type": "boolean"},
        "is_friendly": {"type": "boolean"},
        "level": {"type": "integer", "minimum": 0},
        "race": {"type": "string", "maxLength": 255},
        "background": {"type": "string", "maxLength": 255},
        "alignment": {"type": "integer", "minimum": 0, "maximum": 8},
        "experience_points": {"type": "integer", "minimum": 0},
        "inspiration": {"type": "integer"},
        "proficiency_bonus": {"type": "integer"},
        "passive_wisdom": {"type": "integer"},
        "other_proficiencies_skills": {"type": "string", "maxLength": 4095},
        "armor_class": {"type": "integer"},
        "initiative": {"type": "integer"},
        "speed": {"type": "integer"},
        "max_hp": {"type": "integer", "minimum": 0},
        "cur_hp": {"type": "integer", "minimum": 0},
        "temp_hp": {"type": "integer", "minimum": 0},
        "hit_dice": {"type": "integer", "minimum": 0},
        "death_save_success": {"type": "integer", "minimum": 0},
        "death_save_failure": {"type": "integer", "minimum": 0},
        "attacks": {"type": "string", "maxLength": 4095},
        "equipment": {"type": "string", "maxLength": 4095},
        "traits": {"type": "string", "maxLength": 4095},
        "ideals": {"type": "string", "maxLength": 4095},
        "bonds": {"type": "string", "maxLength": 4095},
        "flaws": {"type": "string", "maxLength": 4095},
        "features_and_traits": {"type": "string", "maxLength": 4095},
        "age": {"type": "integer", "minimum": 0},
        "eyes": {"type": "string", "maxLength": 255},
        "height": {"type": "string", "maxLength": 255},
        "weight": {"type": "string", "maxLength": 255},
        "backstory": {"type": "string", "maxLength": 4095},
        "dark_gifts": {"type": ["string", "null"], "maxLength": 4095},
        "features": {"type": "string", "maxLength": 4095},
        "allies": {"type": ["string", "null"], "maxLength": 4095},
        "treasure": {"type": ["string", "null"], "maxLength": 4095},
        "notes": {"type": ["string", "null"], "maxLength": 4095},

        "abilities": {
            "type": "object",
            "properties": {
                ability: {"type": "integer"}
                for ability in abilities
            },
            "additionalProperties": False,
            "minProperties": len(abilities),
        },
        "saving": {
            "type": "object",
            "properties": {
                ability: {"type": "boolean"}
                for ability in abilities
            },
            "additionalProperties": False,
            "minProperties": len(abilities),
        },
        "skills": {
            "type": "object",
            "properties": {
                skill: {"type": "boolean"}
                for skill in skills
            },
            "additionalProperties": False,
            "minProperties": len(skills),
        },

        "spells": {
            "type": "object",
            "properties": {
                "spellcast_ability": {"type": "string", "maxLength": 127},
                "spell_save_dc": {"type": "integer"},
                "spell_attack_bonus": {"type": "integer"},
                "by_level": {
                    "type": "array",
                    "items": {
                        "type": "object",
                        "properties": {
                            "slots": {"type": "integer", "minimum": 0},
                            "spells": {
                                "type": "array",
                                "items": {
                                    "type": "object",
                                    "properties": {
                                        "name": {"type": "string", "maxLength": 255},
                                        "charges": {"type": "integer", "minimum": 0},
                                    },
                                    "required": ["name"],
                                },
                            },
                        },
                    },
                    "minItems": 11, # 0-x
                    "maxItems": 11,
                },
            },
            "additionalProperties": False,
            "minProperties": 4,
        },
    },
    "additionalProperties": False,
    "minProperties": 43,
}

def create_character_sheet(info, is_npc):
    try:
        validate(instance=info, schema=character_sheet_schema)
    except ValidationError as e:
        logger.warn(f"Got validation error: {e}")
        return None

    if info["is_npc"] != is_npc:
        logger.warn(f"is_npc: Expected {is_npc}, got {info['is_npc']}")
        return None

    with transaction.atomic():
        sheet = app.models.CharacterSheet()
        for attr in same_attrs:
            setattr(sheet, attr, info[attr])
        for ability, val in info["abilities"].items():
            setattr(sheet, f"{ability}_ability", val)
        for ability, val in info["saving"].items():
            setattr(sheet, f"{ability}_saving", val)
        for skill, val in info["skills"].items():
            setattr(sheet, f"{skill}_skill", val)
        sheet.save()

        spellsheet = app.models.CharacterSpellSheet(character=sheet)
        for attr in spells_same_attrs:
            setattr(spellsheet, attr, info["spells"][attr])
        for level, level_info in enumerate(info["spells"]["by_level"]):
            if level == 10:
                level = "x"
            setattr(spellsheet, f"level_{level}_spellslots", level_info["slots"])
        spellsheet.save()

        for level, level_info in enumerate(info["spells"]["by_level"]):
            for spell_info in level_info["spells"]:
                spell = app.models.Spell.objects.filter(name=spell_info["name"])
                if len(spell) != 1:
                    transaction.rollback()
                    logger.warn(f"Unknown spell f{spell_info['name']}")
                    return None
                spell = spell[0]
                spellslot = app.models.CharacterSpellslot(character=spellsheet, spell=spell, charges=spell_info["charges"])
                spellslot.save()

        return sheet

def json_spell(spellslot):
    json = {}
    for attr in spell_same_attrs:
        json[attr] = getattr(spellslot.spell, attr)
    json["charges"] = spellslot.charges
    return json

def json_spells(sheet):
    spellsheet = app.models.CharacterSpellSheet.objects.filter(character=sheet)
    if len(spellsheet) != 1:
        return None
    spellsheet = spellsheet[0]
    json = {}
    for attr in spells_same_attrs:
        json[attr] = getattr(spellsheet, attr)
    json["by_level"] = []
    for level in range(0, 10):
        json["by_level"].append({ "slots": getattr(spellsheet, f"level_{level}_spellslots"), "spells": [] })
    json["by_level"].append({ "slots": spellsheet.level_x_spellslots, "spells": [] })

    spells = app.models.CharacterSpellslot.objects.filter(character=spellsheet)
    for spell in spells:
        json["by_level"][spell.spell.level]["spells"].append(json_spell(spell))
    return json

def json_character_sheet(sheet):
    json = {}
    for attr in same_attrs:
        json[attr] = getattr(sheet, attr)
    json["abilities"] = {}
    json["saving"] = {}
    for ability in abilities:
        json["abilities"][ability] = getattr(sheet, f"{ability}_ability")
        json["saving"][ability] = getattr(sheet, f"{ability}_saving")
    json["skills"] = {}
    for skill in skills:
        json["skills"][ability] = getattr(sheet, f"{skill}_skill")
    json["spells"] = json_spells(sheet)
    return json


